eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -1,132 +1,451 @@
+pfhConstitutionL = lens pfConstitution $ \x y -> x {pfConstitution = y}
 
-instance EraPParams era => NoThunks (Proposals era)
+deriving instance Era era => Eq (PForest PRoot era)
+deriving instance Era era => Ord (PForest PRoot era)
+deriving instance Era era => Generic (PForest PRoot era)
+deriving instance Era era => NoThunks (PForest PRoot era)
+deriving instance Era era => NFData (PForest PRoot era)
+deriving instance Era era => Default (PForest PRoot era)
 
-instance Default (Proposals era) where
-  def = Proposals def
+deriving instance Era era => Eq (PForest PHierarchy era)
+deriving instance Era era => Ord (PForest PHierarchy era)
+deriving instance Era era => Generic (PForest PHierarchy era)
+deriving instance Era era => NoThunks (PForest PHierarchy era)
+deriving instance Era era => NFData (PForest PHierarchy era)
+deriving instance Era era => Default (PForest PHierarchy era)
+
+data Proposals era = Proposals
+  { pOMap :: !(OMap.OMap (GovActionId (EraCrypto era)) (GovActionState era))
+  , pRoots :: !(PForest PRoot era)
+  , pHierarchy :: !(PForest PHierarchy era)
+  }
+  deriving (Eq, Generic, NoThunks, NFData, Default)
+
+instance Era era => Show (Proposals era) where
+  show = flip proposalsShowDebug True
+
+pOMapL :: Lens' (Proposals era) (OMap.OMap (GovActionId (EraCrypto era)) (GovActionState era))
+pOMapL = lens pOMap $ \x y -> x {pOMap = y}
+
+pRootsL :: Lens' (Proposals era) (PForest PRoot era)
+pRootsL = lens pRoots $ \x y -> x {pRoots = y}
+
+pHierarchyL :: Lens' (Proposals era) (PForest PHierarchy era)
+pHierarchyL = lens pHierarchy $ \x y -> x {pHierarchy = y}
+
+instance EraPParams era => ToJSON (Proposals era) where
+  toJSON = toJSON . pOMap
+  toEncoding = toEncoding . pOMap
+
+proposalsAddAction :: forall era. GovActionState era -> Proposals era -> Maybe (Proposals era)
+proposalsAddAction gas ps =
+  -- TODO
+  -- case getLenses $ gas ^. gasActionL of
+  --   Nothing -> Just $ ps & pOMapL %~ (OMap.||> gas) -- TredsuryWithdrawals and InfoAction do not have their own trees
+  --   Just (lenzR, lenzH, parent) -> update lenzR lenzH parent
+  case gas ^. gasActionL of
+    ParameterChange parent _ ->
+      update pfrPParamUpdateL pfhPParamUpdateL parent
+    HardForkInitiation parent _ ->
+      update pfrHardForkL pfhHardForkL parent
+    TreasuryWithdrawals _ -> Just $ ps & pOMapL %~ (OMap.||> gas)
+    NoConfidence parent ->
+      update pfrCommitteeL pfhCommitteeL parent
+    UpdateCommittee parent _ _ _ ->
+      update pfrCommitteeL pfhCommitteeL parent
+    NewConstitution parent _ ->
+      update pfrConstitutionL pfhConstitutionL parent
+    InfoAction -> Just $ ps & pOMapL %~ (OMap.||> gas)
+  where
+    update ::
+      Lens' (PForest PRoot era) (PRoot (GovPurposeId p era)) ->
+      Lens' (PForest PHierarchy era) (PHierarchy (GovPurposeId p era)) ->
+      StrictMaybe (GovPurposeId p era) ->
+      Maybe (Proposals era)
+    update lenzR lenzH prev =
```

Comment:
Out of curiosity, why did you opt for `z` in `lenz` instead of `s` ? Is it to avoid a name collision?

Reply:
Yes! It is to avoid name collision.